add support for axis network speaker

jonoberheide opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

Hi! First-time HA developer here, so (1) adding this support took a lot longer/more learning than the LOC may indicate, and (2) open to feedback on anything I may have done incorrectly.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add support for Axis Network Speakers, exposing the device as a Media Player.

Tested on the Axis C1410 Mk II Network Mini Speaker.

@Kane610 Right now this is just adding the basic support to broadcast audio via the Axis speaker and the Audio API transmit.cgi. There is little/no state tracking, no event updates from the device, etc. I am guessing this would work with any Axis devices that support the Audio API (eg. a camera that also has a built-in speaker), but I've only tested it on the dedicated C1410 speaker device.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

jonoberheide committed:
add support for axis network speaker

Kane610 commented on homeassistant/components/axis/media_player.py:
```diff
@@ -0,0 +1,146 @@
+        output = await ffmpeg_cmd(ffmpeg_manager.binary, media_id)
+
+        uri = "/axis-cgi/audio/transmit.cgi"
+        conf = self.hub.config
+        url = f"{conf.protocol}://{conf.host}:{conf.port}{uri}"
+        auth = httpx.DigestAuth(self.hub.config.username, self.hub.config.password)
+        headers = {"Content-Type": "audio/axis-mulaw-128"}
+
+        client = get_async_client(self.hass, verify_ssl=False)
+        await client.post(url, headers=headers, content=output, auth=auth)
```
Use the `self.hub.api.vapix.request` method, you should probably move everything related to the request into the axis library, keep ffmpeg stuff here
```suggestion
        headers = {"Content-Type": "audio/axis-mulaw-128"}
        await self.hub.api.vapix.request("post", uri, headers=headers, content=output)
```

Kane610 reviewed (changes requested):
This is awesome! Thank you

I would really like to see parts related to transmit.cgi be moved and defined in the axis library instead. Keeps everything together. 

Im also missing a check somewhere that validates that the device has audio out.

And once all of that is fixed Im missing tests.

I will be busy for the next two weeks but am available for support after that.

Have you referenced other integrations using ffmpeg if there are some common pattern recommended to be used?

jonoberheide committed:
Update homeassistant/components/axis/media_player.py

Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>

jonoberheide commented:
> I would really like to see parts related to transmit.cgi be moved and defined in the axis library instead. Keeps everything together.

Makes sense. I figured there was some logical subset to have in the axis lib. I kept everything in HA-land to start just for simplicity.

> Im also missing a check somewhere that validates that the device has audio out.

I will poke at this. I have a couple Axis cameras and a network relay so hopefully those should be good devices to validate against.
 
> And once all of that is fixed Im missing tests.

HA test suite is new territory, but I'll give it a shot.

> Have you referenced other integrations using ffmpeg if there are some common pattern recommended to be used?

Yes, I modeled it roughly off a few other components that use ffmpeg for audio transcoding like tts/uiprotect/etc. Seems ffmpeg_manager is best practice in HA-land, while some components have the ffmpeg logic in the non-HA helper library. The latter could make sense here in terms of standalone usage/testing from the axis library, but ffmpeg_manager is nice for path discovery and such. One middle-ground might be to use ffmpeg_manager and pass the path into the axis library and have the rest of the logic in there. I think this is roughly what uiprotect does IIRC.

Some components use temp files with ffmpeg, but I wanted to keep things off disk with subprocess pipes. Also, in theory, ffmpeg itself could directly stream the resulting audio to the transmit.cgi URL, but keeping things native/async with vapix/httpx is better IMO.

Kane610 commented:
Lets take the discussion on ffmpeg once Im back. I think its good to provide s ffmpeg client into the library similar to how the httpx client is provided. Because then all axis details are kept in the library which is what we want.

I have rewritten most tests for axis this summer so they are att least a bit more modern

jonoberheide committed:
- use the new axis library api
- move ffmpeg logic into the axis library
- add volume/mute support
- implement async_update to get latest device state
- only add a media player entity if both audio APIs are supported

jonoberheide commented:
New changes:
- use the new axis library api (from https://github.com/Kane610/axis/pull/521)
- move ffmpeg logic into the axis library
- add volume/mute support
- implement async_update to get latest device state
- only add a media player entity if both audio APIs are supported

jonoberheide committed:
catch errors if the media url can't be resolved/converted for some reason

Joepkke commented:
Hi Guys,

Great work! Is it also possible to add a function for push to speak. I've a axis M1075 camera with 2 way audio.
